arangod/IResearch/IResearchAnalyzerFeature.cpp

```diff
@@ -1644,6 +1717,227 @@ Result IResearchAnalyzerFeature::emplace(
+    res = trx.commitAsync().get();
+    if (res.ok()) {
+      invalidate(vocbase);
+    }
+    return res;
+  } else {
+    // as in single analyzer case. First mark for deletion (but all at once)
+    auto stringRev = std::to_string(analyzerModificationTrx->buildingRevision());
+    std::string aql("FOR u IN ");
+    aql.append(arangodb::StaticStrings::AnalyzersCollection).append(" UPDATE u.")
+       .append(arangodb::StaticStrings::KeyString).append(" WITH { ")
+       .append(arangodb::StaticStrings::AnalyzersDeletedRevision)
+       .append(": ").append(stringRev).append("} IN ")
+       .append(arangodb::StaticStrings::AnalyzersCollection);
+
+    auto ctx = transaction::StandaloneContext::Create(vocbase);
+    SingleCollectionTransaction trx(ctx, arangodb::StaticStrings::AnalyzersCollection,
+                                    AccessMode::Type::EXCLUSIVE);
+
+    auto res = trx.begin();
+    if (res.fail()) {
+      return res;
+    }
+    arangodb::aql::Query query(ctx, arangodb::aql::QueryString(aql), nullptr, nullptr);
+    aql::QueryResult queryResult = query.executeSync();
+    if (queryResult.fail()) {
+      return queryResult.result;
+    }
+    res = trx.commitAsync().get();
+    if (!res.ok()) {
+      return res;
+    }
+    res = analyzerModificationTrx->commit();
+    if (res.fail()) {
+      return res;
+    }
+    // now let`s do cleanup
+    SingleCollectionTransaction truncateTrx(ctx, arangodb::StaticStrings::AnalyzersCollection,
+                                            AccessMode::Type::EXCLUSIVE);
+
+    res = truncateTrx.begin();
+
+    if (res.ok()) {
+      OperationOptions options;
+      truncateTrx.truncateAsync(arangodb::StaticStrings::AnalyzersCollection, options).get();
+      res = truncateTrx.commitAsync().get();
+    }
+    if (res.fail()) {
+      // failed cleanup is not critical problem. just log it
+      LOG_TOPIC("70a8c", WARN, iresearch::TOPIC) << " Failed to finalize analyzer truncation"
+        << " Error Code:" << res.errorNumber() << " Error:" << res.errorMessage();
+    }
+    invalidate(vocbase);
+    return {};
+  }
+}
+
+Result IResearchAnalyzerFeature::bulkEmplace(TRI_vocbase_t& vocbase,
+                                             VPackSlice const dumpedAnalzyzers) {
+  TRI_ASSERT(dumpedAnalzyzers.isArray());
+  TRI_ASSERT(!dumpedAnalzyzers.isEmptyArray());
+  auto transaction = createAnalyzerModificationTransaction(server(), vocbase.name());
+  if (transaction) {
+    auto startRes = transaction->start();
+    if (startRes.fail()) {
+      return startRes;
+    }
+  }
+  try {
+    WriteMutex mutex(_mutex);
+    SCOPED_LOCK(mutex);
+
+    if (transaction) {
+      auto cleanupResult = cleanupAnalyzersCollection(vocbase.name(),
+                                                      transaction->buildingRevision());
+      if (cleanupResult.fail()) {
+        return cleanupResult;
+      }
+    }
+    auto res = loadAnalyzers(vocbase.name());
```

Comment:
mmmh, as far as I can see `loadAnalyzers()` also acquires the write-mutex on `_mutex`: https://github.com/arangodb/arangodb/pull/12059/files#diff-2eae74915725dafeff75a70e1882fafcL1981
So the mutex would be acquired recursively. Is that safe to do?

Reply:
It is safe as mutex used supports recursive locking. However I guess it is a good point to move lock lower to allow access to analyzers while  cleanup operation runs (as cleanup don`t touch already cached analyzers it is safe)